lib/elixir/pages/anti-patterns/code-anti-patterns.md

```diff
@@ -0,0 +1,39 @@
+# Code-related anti-patterns
+
+This document outlines anti-patterns related to your code and particular Elixir idioms and features.
+
+## Comments
+
+TODO.
+
+## Long parameter list
+
+TODO.
+
+## Complex branching
+
+TODO.
+
+## Complex else clauses in with
+
+TODO.
+
+## Complex extractions in clauses
+
+TODO.
+
+## Dynamic map fields access
+
+TODO.
+
+## Dynamic atom creation
+
+TODO.
+
+## Namespace violation
+
+TODO. (note: this is defining modules not in your namespace)
+
+## Speculative Generality
+
+TODO. (note: it is Speculative Assumptions but renamed)
```

Comment:
We decided to merge these two.

Reply:
@josevalim I'm a bit confused about this merge. Although both smells involve speculations, they address different situations in my opinion.

While **Speculative Assumptions** "speculates" that the input formats of a function will always be compatible with its behavior, thus preventing code from breaking when it should, **Speculative Generality** "speculates" that the code may evolve in a particular direction in the future, leading to the addition of features to make it more flexible for changes.

Note that in both cases, we have speculations, but they pertain to different aspects and have impacts that are also difficult to compare. **Speculative Assumptions** can generate code with incorrect behaviors/returns, meaning it can induce bugs (although it's not considered one itself). On the other hand, **Speculative Generality** only makes the code more verbose and may compromise its readability but does not cause unintended behaviors in the code.

Wouldn't it be better to reconsider this merge and keep these two smells separate?

Reply:
Thank you. We will keep it as "Speculative Assumptions" then.

The blog post it came from calls it "Assertive code", I wonder if we could also call it something else, like "Loose code". But it is fine either way. :)

Reply:
I think this is sort of multiple versions of YAGNI to be fair, or "premature abstraction".

Reply:
The original example though is more about accidentally dealing with different types of inputs rather than being intentional about it (which premature abstraction is about).